File: App/Backend/src/services/auth_service.py
from datetime import datetime, timedelta
from uuid import uuid4

from sqlalchemy.orm import Session

from Backend.src.core.otp_utils import (
    generate_otp,
    get_otp_expiry,
    hash_otp,
    verify_otp,
)
from Backend.src.core.security import (
    create_access_token,
    create_refresh_token,
    hash_password,
    hash_refresh_token,
    verify_password,
)
from Backend.src.repositories.student_repository import StudentRepository
from Backend.src.repositories.user_repository import UserRepository
from Backend.src.services.email_service import send_otp_email

# Minimum seconds between resend requests
OTP_RESEND_COOLDOWN_SECONDS = 60


def generate_uid(db: Session) -> str:
    last_user = UserRepository.get_last_user(db)
    if not last_user:
        return "USR001"

    number = int(last_user.uid.replace("USR", ""))
    return f"USR{number + 1:03d}"


def register_user(
    db: Session,
    username: str,
    email: str,
    recovery_email: str | None,
    password: str,
    name: str,
):
    """
    Publicly accessible student registration.
    - Creates User with role='student'
    - Auto-creates Student profile with provided name
    - Sends email verification OTP via Brevo SMTP
    """
    existing_email = UserRepository.get_by_email(db, email)
    if existing_email:
        raise ValueError("User with this email already exists")

    existing_username = UserRepository.get_by_username(db, username)
    if existing_username:
        raise ValueError("Username already exists")

    if recovery_email and recovery_email == email:
        raise ValueError("Recovery email cannot be the same as primary email")

    uid = generate_uid(db)

    user_data = {
        "uid": uid,
        "username": username,
        "email": email,
        "recovery_email": recovery_email,
        "password_hash": hash_password(password),
        "role": "student",
        "email_verified": False,
        "recovery_email_verified": False,
        "is_active": True,
        "deactivated_at": None,
    }

    user = UserRepository.create(db, user_data)
    StudentRepository.create(db, {"uid": uid, "name": name})

    # Send verification OTP via email
    otp = create_otp(db, uid, "email_verification")
    send_otp_email(
        to_email=email,
        otp=otp,
        purpose="email_verification",
        username=username
    )

    return {
        "uid": user.uid,
        "username": user.username,
        "email": user.email,
        "role": user.role,
        "email_verified": user.email_verified,
        "is_active": user.is_active,
        "message": "Registration successful. A verification OTP has been sent to your email."
    }


def authenticate_user(
    db: Session,
    email: str,
    password: str,
    remember_me: bool = False
):
    email = email.strip().lower()
    user = UserRepository.get_by_email(db, email)

    if not user:
        return None

    if not verify_password(password, user.password_hash):
        return None

    if not user.is_active:
        raise ValueError("Your account is deactivated")

    access_token = create_access_token(
        uid=user.uid,
        role=user.role
    )

    session = create_user_session(db, user.uid, remember_me)

    return {
        "uid": user.uid,
        "username": user.username,
        "email": user.email,
        "role": user.role,
        "email_verified": user.email_verified,
        "access_token": access_token,
        "refresh_token": session["refresh_token"],
        "session_id": session["session_id"],
        "token_type": "bearer"
    }


def create_otp(
    db: Session,
    uid: str,
    purpose: str
):
    allowed_purposes = [
        "email_verification",
        "password_reset",
        "recovery_email_verification"
    ]

    if purpose not in allowed_purposes:
        raise ValueError("Invalid OTP purpose")

    user = UserRepository.get_by_uid(db, uid)
    if not user:
        raise ValueError("User not found")

    # Invalidate any existing active OTPs for this purpose
    UserRepository.invalidate_existing_otps(db, uid, purpose)

    otp = generate_otp()
    UserRepository.create_otp_record(
        db=db,
        uid=uid,
        otp_hash=hash_otp(otp),
        purpose=purpose,
        expires_at=get_otp_expiry()
    )

    return otp


def verify_user_otp(
    db: Session,
    uid: str,
    purpose: str,
    otp: str
):
    otp_record = UserRepository.get_active_otp(db, uid, purpose)

    if not otp_record:
        raise ValueError("OTP not found")

    if otp_record.expires_at < datetime.utcnow():
        raise ValueError("OTP has expired")

    if otp_record.attempts >= 5:
        raise ValueError("Maximum OTP attempts exceeded")

    if not verify_otp(otp, otp_record.otp_hash):
        UserRepository.increment_otp_attempts(db, otp_record)
        raise ValueError("Invalid OTP")

    UserRepository.mark_otp_used(db, otp_record)
    return True


def request_email_verification(
    db: Session,
    email: str
):
    email = email.strip().lower()
    user = UserRepository.get_by_email(db, email)

    if not user:
        raise ValueError("User not found")

    if user.email_verified:
        raise ValueError("Email is already verified")

    otp = create_otp(db, user.uid, "email_verification")

    email_sent = send_otp_email(
        to_email=email,
        otp=otp,
        purpose="email_verification",
        username=user.username
    )

    if not email_sent:
        raise ValueError(
            "OTP was generated, but the verification email could not be sent."
        )

    return True


def resend_otp(
    db: Session,
    email: str,
    purpose: str
):
    allowed_purposes = [
        "email_verification",
        "password_reset",
        "recovery_email_verification"
    ]

    if purpose not in allowed_purposes:
        raise ValueError("Invalid OTP purpose")

    email = email.strip().lower()
    user = UserRepository.get_by_email(db, email)

    if not user:
        raise ValueError("User not found")

    if purpose == "email_verification" and user.email_verified:
        raise ValueError("Email is already verified")

    # Check cooldown
    last_otp = UserRepository.get_latest_otp(db, user.uid, purpose)
    if last_otp:
        elapsed = (datetime.utcnow() - last_otp.created_at).total_seconds()
        if elapsed < OTP_RESEND_COOLDOWN_SECONDS:
            remaining = int(OTP_RESEND_COOLDOWN_SECONDS - elapsed)
            raise ValueError(
                f"Please wait {remaining} seconds before requesting a new OTP."
            )

    otp = create_otp(db, user.uid, purpose)

    send_otp_email(
        to_email=email,
        otp=otp,
        purpose=purpose,
        username=user.username
    )

    return True


def verify_email(
    db: Session,
    email: str,
    otp: str
):
    email = email.strip().lower()
    user = UserRepository.get_by_email(db, email)

    if not user:
        raise ValueError("User not found")

    if user.email_verified:
        raise ValueError("Email is already verified")

    verify_user_otp(db, user.uid, "email_verification", otp)
    UserRepository.update(db, user, {"email_verified": True})
    return True


def request_password_reset(
    db: Session,
    email: str
):
    email = email.strip().lower()
    user = UserRepository.get_by_email(db, email)

    if not user:
        raise ValueError("User not found")

    if not user.is_active:
        raise ValueError("User account is deactivated")

    otp = create_otp(db, user.uid, "password_reset")

    send_otp_email(
        to_email=email,
        otp=otp,
        purpose="password_reset",
        username=user.username
    )

    return True


def reset_user_password(
    db: Session,
    email: str,
    otp: str,
    new_password: str
):
    email = email.strip().lower()
    user = UserRepository.get_by_email(db, email)

    if not user:
        raise ValueError("User not found")

    verify_user_otp(db, user.uid, "password_reset", otp)

    if verify_password(new_password, user.password_hash):
        raise ValueError("New password cannot be the same as old password")

    UserRepository.update(db, user, {"password_hash": hash_password(new_password)})
    return True


def create_user_session(
    db: Session,
    uid: str,
    remember_me: bool = False
):
    refresh_token = create_refresh_token()
    expires_at = datetime.utcnow() + timedelta(
        days=30 if remember_me else 7
    )
    session_id = str(uuid4())

    session = UserRepository.create_session(
        db=db,
        session_id=session_id,
        uid=uid,
        refresh_token_hash=hash_refresh_token(refresh_token),
        expires_at=expires_at
    )

    return {
        "session_id": session.session_id,
        "refresh_token": refresh_token,
        "expires_at": session.expires_at
    }


def refresh_access_token(
    db: Session,
    refresh_token: str
):
    hashed_token = hash_refresh_token(refresh_token)
    session = UserRepository.get_session_by_hash(db, hashed_token)

    if not session:
        raise ValueError("Invalid session")

    if session.expires_at < datetime.utcnow():
        raise ValueError("Session expired")

    user = UserRepository.get_by_uid(db, session.uid)
    if not user:
        raise ValueError("User not found")

    if not user.is_active:
        raise ValueError("User account is deactivated")

    UserRepository.update_session_usage(db, session)
    return create_access_token(uid=user.uid, role=user.role)


def logout_user(
    db: Session,
    session_id: str,
    uid: str
):
    session = UserRepository.get_active_session_by_id(db, session_id, uid)
    if not session:
        raise ValueError("Active session not found")

    UserRepository.revoke_session(db, session)
    return True

File: App(SpringBoot)/src/main/java/com/Project/LearningManagementSystem/service/AuthService.java
package com.Project.LearningManagementSystem.service;

import com.Project.LearningManagementSystem.dto.AuthDtos.LoginRequest;
import com.Project.LearningManagementSystem.dto.AuthDtos.RegisterRequest;
import com.Project.LearningManagementSystem.entity.OtpVerification;
import com.Project.LearningManagementSystem.entity.Student;
import com.Project.LearningManagementSystem.entity.User;
import com.Project.LearningManagementSystem.entity.UserSession;
import com.Project.LearningManagementSystem.exception.BadRequestException;
import com.Project.LearningManagementSystem.exception.UnauthorizedException;
import com.Project.LearningManagementSystem.repository.OtpVerificationRepository;
import com.Project.LearningManagementSystem.repository.StudentRepository;
import com.Project.LearningManagementSystem.repository.UserRepository;
import com.Project.LearningManagementSystem.repository.UserSessionRepository;
import com.Project.LearningManagementSystem.security.JwtUtils;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Random;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final StudentRepository studentRepository;
    private final OtpVerificationRepository otpRepository;
    private final UserSessionRepository sessionRepository;
    private final EmailService emailService;
    private final PasswordEncoder passwordEncoder;
    private final JwtUtils jwtUtils;

    private static final int OTP_EXPIRY_MINUTES = 2;
    private static final int OTP_RESEND_COOLDOWN_SECONDS = 60;

    public static String hashToken(String token) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(token.getBytes(java.nio.charset.StandardCharsets.UTF_8));
            StringBuilder hexString = new StringBuilder();
            for (byte b : hash) {
                String hex = Integer.toHexString(0xff & b);
                if (hex.length() == 1) hexString.append('0');
                hexString.append(hex);
            }
            return hexString.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

    public synchronized String generateUid() {
        List<User> users = userRepository.findAll();
        if (users.isEmpty()) {
            return "USR001";
        }
        int maxNum = users.stream()
            .map(User::getUid)
            .filter(uid -> uid != null && uid.startsWith("USR"))
            .mapToInt(uid -> {
                try {
                    return Integer.parseInt(uid.substring(3));
                } catch (NumberFormatException e) {
                    return 0;
                }
            })
            .max()
            .orElse(0);

        return String.format("USR%03d", maxNum + 1);
    }

    @Transactional
    public Map<String, Object> registerUser(RegisterRequest request) {
        String email = request.getEmail().trim().toLowerCase();
        if (userRepository.existsByEmail(email)) {
            throw new BadRequestException("User with this email already exists");
        }

        if (request.getRecovery_email() != null && request.getRecovery_email().trim().equalsIgnoreCase(email)) {
            throw new BadRequestException("Recovery email cannot be the same as primary email");
        }

        String uid = generateUid();

        User user = new User();
        user.setUid(uid);
        user.setUsername(request.getUsername().trim());
        user.setEmail(email);
        user.setRecoveryEmail(request.getRecovery_email());
        user.setPasswordHash(passwordEncoder.encode(request.getPassword()));
        user.setRole("student");
        user.setEmailVerified(false);
        user.setRecoveryEmailVerified(false);
        user.setActive(true);
        userRepository.save(user);

        Student student = new Student();
        student.setUid(uid);
        student.setName(request.getName().trim());
        studentRepository.save(student);

        String otp = createOtp(uid, "email_verification");
        emailService.sendOtpEmail(email, otp, "email_verification", user.getUsername());

        Map<String, Object> response = new HashMap<>();
        response.put("uid", user.getUid());
        response.put("username", user.getUsername());
        response.put("email", user.getEmail());
        response.put("role", user.getRole());
        response.put("email_verified", user.isEmailVerified());
        response.put("is_active", user.isActive());
        response.put("message", "Registration successful. A verification OTP has been sent to your email.");
        return response;
    }

    @Transactional
    public Map<String, Object> authenticateUser(LoginRequest request) {
        String email = request.getEmail().trim().toLowerCase();
        User user = userRepository.findByEmail(email)
            .orElseThrow(() -> new UnauthorizedException("Invalid email or password"));

        if (!passwordEncoder.matches(request.getPassword(), user.getPasswordHash())) {
            throw new UnauthorizedException("Invalid email or password");
        }

        if (!user.isActive()) {
            throw new BadRequestException("Your account is deactivated");
        }

        String accessToken = jwtUtils.generateToken(user.getUid(), user.getRole());
        String refreshToken = UUID.randomUUID().toString();
        String sessionId = UUID.randomUUID().toString();

        UserSession session = new UserSession();
        session.setSessionId(sessionId);
        session.setUid(user.getUid());
        session.setRefreshTokenHash(hashToken(refreshToken));
        session.setExpiresAt(LocalDateTime.now().plusDays(request.isRemember_me() ? 30 : 7));
        session.setCreatedAt(LocalDateTime.now());
        session.setLastUsedAt(LocalDateTime.now());
        sessionRepository.save(session);

        Map<String, Object> result = new HashMap<>();
        result.put("uid", user.getUid());
        result.put("username", user.getUsername());
        result.put("email", user.getEmail());
        result.put("role", user.getRole());
        result.put("email_verified", user.isEmailVerified());
        result.put("access_token", accessToken);
        result.put("refresh_token", refreshToken);
        result.put("session_id", sessionId);
        result.put("token_type", "bearer");
        return result;
    }

    @Transactional
    public String createOtp(String uid, String purpose) {
        String otp = String.format("%06d", new Random().nextInt(1_000_000));

        OtpVerification record = new OtpVerification();
        record.setUid(uid);
        record.setOtpHash(passwordEncoder.encode(otp));
        record.setPurpose(purpose);
        record.setExpiresAt(LocalDateTime.now().plusMinutes(OTP_EXPIRY_MINUTES));
        record.setCreatedAt(LocalDateTime.now());
        record.setAttempts(0);
        record.setUsed(false);
        otpRepository.save(record);

        return otp;
    }

    public void verifyUserOtp(String uid, String purpose, String otp) {
        OtpVerification record = otpRepository.findTopByUidAndPurposeOrderByCreatedAtDesc(uid, purpose)
            .orElseThrow(() -> new BadRequestException("OTP not found"));

        if (record.isUsed()) {
            throw new BadRequestException("OTP has already been used");
        }
        if (record.getExpiresAt().isBefore(LocalDateTime.now())) {
            throw new BadRequestException("OTP has expired");
        }
        if (record.getAttempts() >= 5) {
            throw new BadRequestException("Maximum OTP attempts exceeded");
        }

        if (!passwordEncoder.matches(otp, record.getOtpHash())) {
            record.setAttempts(record.getAttempts() + 1);
            otpRepository.save(record);
            throw new BadRequestException("Invalid OTP");
        }

        record.setUsed(true);
        otpRepository.save(record);
    }

    @Transactional
    public void requestEmailVerification(String email) {
        email = email.trim().toLowerCase();
        User user = userRepository.findByEmail(email)
            .orElseThrow(() -> new BadRequestException("User not found"));

        if (user.isEmailVerified()) {
            throw new BadRequestException("Email is already verified");
        }

        String otp = createOtp(user.getUid(), "email_verification");
        emailService.sendOtpEmail(email, otp, "email_verification", user.getUsername());
    }

    @Transactional
    public void resendOtp(String email, String purpose) {
        email = email.trim().toLowerCase();
        User user = userRepository.findByEmail(email)
            .orElseThrow(() -> new BadRequestException("User not found"));

        if ("email_verification".equals(purpose) && user.isEmailVerified()) {
            throw new BadRequestException("Email is already verified");
        }

        Optional<OtpVerification> lastOtp = otpRepository.findTopByUidAndPurposeOrderByCreatedAtDesc(user.getUid(), purpose);
        if (lastOtp.isPresent()) {
            long secondsElapsed = java.time.Duration.between(lastOtp.get().getCreatedAt(), LocalDateTime.now()).getSeconds();
            if (secondsElapsed < OTP_RESEND_COOLDOWN_SECONDS) {
                long remaining = OTP_RESEND_COOLDOWN_SECONDS - secondsElapsed;
                throw new BadRequestException("Please wait " + remaining + " seconds before requesting a new OTP.");
            }
        }

        String otp = createOtp(user.getUid(), purpose);
        emailService.sendOtpEmail(email, otp, purpose, user.getUsername());
    }

    @Transactional
    public void verifyEmail(String email, String otp) {
        email = email.trim().toLowerCase();
        User user = userRepository.findByEmail(email)
            .orElseThrow(() -> new BadRequestException("User not found"));

        if (user.isEmailVerified()) {
            throw new BadRequestException("Email is already verified");
        }

        verifyUserOtp(user.getUid(), "email_verification", otp);
        user.setEmailVerified(true);
        userRepository.save(user);
    }

    @Transactional
    public void requestPasswordReset(String email) {
        email = email.trim().toLowerCase();
        User user = userRepository.findByEmail(email)
            .orElseThrow(() -> new BadRequestException("User not found"));

        if (!user.isActive()) {
            throw new BadRequestException("User account is deactivated");
        }

        String otp = createOtp(user.getUid(), "password_reset");
        emailService.sendOtpEmail(email, otp, "password_reset", user.getUsername());
    }

    @Transactional
    public void resetPassword(String email, String otp, String newPassword) {
        email = email.trim().toLowerCase();
        User user = userRepository.findByEmail(email)
            .orElseThrow(() -> new BadRequestException("User not found"));

        verifyUserOtp(user.getUid(), "password_reset", otp);

        if (passwordEncoder.matches(newPassword, user.getPasswordHash())) {
            throw new BadRequestException("New password cannot be the same as old password");
        }

        user.setPasswordHash(passwordEncoder.encode(newPassword));
        userRepository.save(user);
    }

    @Transactional
    public String refreshAccessToken(String refreshToken) {
        String hashedToken = hashToken(refreshToken);
        UserSession session = sessionRepository.findByRefreshTokenHash(hashedToken)
            .orElseThrow(() -> new UnauthorizedException("Invalid session"));

        if (session.isRevoked() || session.getExpiresAt().isBefore(LocalDateTime.now())) {
            throw new UnauthorizedException("Session expired");
        }

        User user = userRepository.findById(session.getUid())
            .orElseThrow(() -> new BadRequestException("User not found"));

        if (!user.isActive()) {
            throw new BadRequestException("User account is deactivated");
        }

        session.setLastUsedAt(LocalDateTime.now());
        sessionRepository.save(session);

        return jwtUtils.generateToken(user.getUid(), user.getRole());
    }

    @Transactional
    public void logoutUser(String sessionId, String uid) {
        Optional<UserSession> sessionOpt = sessionRepository.findBySessionIdAndUid(sessionId, uid);
        sessionOpt.ifPresent(session -> {
            session.setRevoked(true);
            sessionRepository.save(session);
        });
    }
}
